src/server/ua_subscription.h

```diff
@@ -54,6 +54,43 @@ typedef struct UA_EventNotification {
     /* EventFilterResult currently isn't being used
     UA_EventFilterResult result; */
 } UA_EventNotification;
+
+#ifdef UA_ENABLE_SUBSCRIPTIONS_ALARMS_CONDITIONS
+typedef struct UA_SpecificCallbacks_Data {
+    UA_TwoStateVariableChangeCallback enteringEnabledStateCallback;
+    UA_TwoStateVariableChangeCallback enteringAckedStateCallback;
+    UA_Boolean ackedRemoveBranch;
+    UA_TwoStateVariableChangeCallback enteringConfirmedStateCallback;
+    UA_Boolean confirmedRemoveBranch;
+    UA_TwoStateVariableChangeCallback enteringActiveStateCallback;
+} UA_SpecificCallbacks_Data;
+
+typedef struct UA_LastSverity_Data {
+    UA_UInt16 lastSeverity;
+    UA_DateTime sourceTimeStamp;
+} UA_LastSverity_Data;
+
+typedef struct UA_ConditionBranch_nodeListElement {
+    LIST_ENTRY(UA_ConditionBranch_nodeListElement) listEntry;
+    UA_NodeId* conditionBranchId;
+    UA_ByteString lastEventId;
+} UA_ConditionBranch_nodeListElement;
+
+typedef struct UA_Condition_nodeListElement {
+    LIST_ENTRY(UA_Condition_nodeListElement) listEntry;
+    LIST_HEAD(conditionbranchlisthead, UA_ConditionBranch_nodeListElement) conditionBranchHead;
+    UA_NodeId conditionId;
```

Comment:
All of these structures have a nodeid and are represented by a node in the information model.
Can we put a pointer to this structure into the node "context"?
This makes lookup more standard.

And we don't have to loop over long linked lists for every access.

Reply:
good idea, I think it would be possible to do that :)

Reply:
@jpfr, actually it would've been a good Idea, if no other application used the context of those nodes. When users would like to attach application specific data to the node context, then, in my opinion, it would not be correct to use up the context for internal logic.

Reply:
You are right. I agree.

Reply:
@jpfr, do you think that using an extra internal context, and leaving the normal context as is, would be a good Idea? I don't know how difficult it is to add such feature to the current implementation, but that would be a nice solution for the linked list looping problem in the Alarms and Conditions implementation (and maybe for other problems as well).

https://github.com/open62541/open62541/pull/2083#discussion_r241076137
for example, a quick solution for that problem would look like this:
(as you see, there are a lot of looping through lists, and that's a real pitfall)

1) In A&C Module:
```C
static void
setIsCallerAC(UA_Server *server,
              const UA_NodeId *condition,
              const UA_NodeId *conditionSource,
              UA_Boolean isCallerAC) {
    /* Get conditionSource Entry */
    UA_ConditionSource_nodeListElement *conditionSourceEntryTmp;
    LIST_FOREACH(conditionSourceEntryTmp, &server->headConditionSource, listEntry) {
        if(!UA_NodeId_equal(&conditionSourceEntryTmp->conditionSourceId, conditionSource))
            continue;
        /* Get Condition Entry */
        UA_Condition_nodeListElement *conditionEntryTmp;
        LIST_FOREACH(conditionEntryTmp, &conditionSourceEntryTmp->conditionHead, listEntry) {
            if(UA_NodeId_equal(&conditionEntryTmp->conditionId, condition)) {// found ConditionId -> branch == NULL
                UA_ConditionBranch_nodeListElement *conditionBranchEntryTmp;
                LIST_FOREACH(conditionBranchEntryTmp, &conditionEntryTmp->conditionBranchHead, listEntry) {
                    if(conditionBranchEntryTmp->conditionBranchId == NULL) {
                        conditionBranchEntryTmp->isCallerAC = isCallerAC;
                        return;
                    }
                }
            }
            else {
                UA_ConditionBranch_nodeListElement *conditionBranchEntryTmp;
                LIST_FOREACH(conditionBranchEntryTmp, &conditionEntryTmp->conditionBranchHead, listEntry) {
                    if(UA_NodeId_equal(conditionBranchEntryTmp->conditionBranchId, condition)) {
                        conditionBranchEntryTmp->isCallerAC = isCallerAC;
                        return;
                    }
                }
            }
        }
    }

    UA_LOG_ERROR(server->config.logger, UA_LOGCATEGORY_USERLAND, "Entry not found in list!");
}

UA_Boolean
isConditionOrBranch(UA_Server *server,
                    const UA_NodeId *condition,
                    const UA_NodeId *conditionSource,
                    UA_Boolean *isCallerAC) {
    /* Get conditionSource Entry */
    UA_ConditionSource_nodeListElement *conditionSourceEntryTmp;
    LIST_FOREACH(conditionSourceEntryTmp, &server->headConditionSource, listEntry) {
        if(!UA_NodeId_equal(&conditionSourceEntryTmp->conditionSourceId, conditionSource))
            continue;
        /* Get Condition Entry */
        UA_Condition_nodeListElement *conditionEntryTmp;
        LIST_FOREACH(conditionEntryTmp, &conditionSourceEntryTmp->conditionHead, listEntry) {
            if(UA_NodeId_equal(&conditionEntryTmp->conditionId, condition)) {// found ConditionId -> branch == NULL
                UA_ConditionBranch_nodeListElement *conditionBranchEntryTmp;
                    LIST_FOREACH(conditionBranchEntryTmp, &conditionEntryTmp->conditionBranchHead, listEntry) {
                        if(conditionBranchEntryTmp->conditionBranchId == NULL) {
                          *isCallerAC = conditionBranchEntryTmp->isCallerAC;
                          return true;
                        }
                    }
            }
            else {
                UA_ConditionBranch_nodeListElement *conditionBranchEntryTmp;
                LIST_FOREACH(conditionBranchEntryTmp, &conditionEntryTmp->conditionBranchHead, listEntry) {
                    if(UA_NodeId_equal(conditionBranchEntryTmp->conditionBranchId, condition)) {
                        *isCallerAC = conditionBranchEntryTmp->isCallerAC;
                        return true;
                    }
                }
            }
        }
    }

    return false;
}
```
```C
UA_StatusCode
UA_Server_triggerConditionEvent(UA_Server *server, const UA_NodeId condition,
                                const UA_NodeId conditionSource, UA_ByteString *outEventId){
    UA_StatusCode retval = UA_STATUSCODE_GOOD;
    UA_ByteString eventId = UA_BYTESTRING_NULL;
    UA_QualifiedName enabledStateField = UA_QUALIFIEDNAME(0, CONDITION_FIELD_ENABLEDSTATE);


    /* Check if enabled */
    if(isTwoStateVariableInTrueState(server, &condition, &enabledStateField)) {
        setIsCallerAC(server, &condition, &conditionSource, true);
        /* Trigger the event for Condition*/
        retval = UA_Server_triggerEvent(server, condition, conditionSource,
                                        &eventId, false); //Condition Nodes should not be deleted after triggering the event
        CONDITION_ASSERT_RETURN_RETVAL(retval, "Triggering condition event failed",);

        setIsCallerAC(server, &condition, &conditionSource, false);
[...]
```

2) In Events Module:
```C
UA_StatusCode
UA_Server_triggerEvent(UA_Server *server, const UA_NodeId eventNodeId, const UA_NodeId origin,
                       UA_ByteString *outEventId, const UA_Boolean deleteEventNode) {

#ifdef UA_ENABLE_SUBSCRIPTIONS_ALARMS_CONDITIONS
    UA_Boolean isCallerAC = false;
    if(isConditionOrBranch(server, &eventNodeId, &origin, &isCallerAC)) {
        if(!isCallerAC) {
          UA_LOG_WARNING(server->config.logger, UA_LOGCATEGORY_SERVER,
                                 "Condition Events: Please use A&C API to trigger Condition Events",
                                  UA_STATUSCODE_BADINVALIDARGUMENT);
          return UA_STATUSCODE_BADINVALIDARGUMENT;
        }
    }
#endif /*UA_ENABLE_SUBSCRIPTIONS_ALARMS_CONDITIONS*/
[...]
```